components/brave_shields/browser/ad_block_service.cc

```diff
@@ -56,21 +56,34 @@ AdBlockService::SourceProviderObserver::~SourceProviderObserver() {
 }
 
 void AdBlockService::SourceProviderObserver::OnChanged() {
-  auto filter_set = std::make_unique<rust::Box<adblock::FilterSet>>(
-      adblock::new_filter_set());
-  rust::Box<adblock::FilterSet>* filter_set_ptr = filter_set.get();
-  auto on_loaded_cb =
-      base::BindOnce(&AdBlockService::SourceProviderObserver::OnFilterSetLoaded,
-                     weak_factory_.GetWeakPtr(), std::move(filter_set));
+  auto on_loaded_cb = base::BindOnce(
+      &AdBlockService::SourceProviderObserver::OnFilterSetCallbackLoaded,
+      weak_factory_.GetWeakPtr());
   if (is_filter_provider_manager_) {
     static_cast<AdBlockFiltersProviderManager*>(filters_provider_.get())
         ->LoadFilterSetForEngine(adblock_engine_->IsDefaultEngine(),
-                                 filter_set_ptr, std::move(on_loaded_cb));
+                                 std::move(on_loaded_cb));
   } else {
-    filters_provider_->LoadFilterSet(filter_set_ptr, std::move(on_loaded_cb));
+    filters_provider_->LoadFilterSet(std::move(on_loaded_cb));
   }
 }
 
+void AdBlockService::SourceProviderObserver::OnFilterSetCallbackLoaded(
+    base::OnceCallback<void(rust::Box<adblock::FilterSet>*)> cb) {
+  task_runner_->PostTaskAndReplyWithResult(
+      FROM_HERE,
+      base::BindOnce(
```

Comment:
in a follow-up I think we should change the callback signature instead of adding another layer to the callbacks